src/request.rs

```diff
@@ -118,70 +127,86 @@ impl Request {
 
     /// Read a raw HTTP request from `reader` and create an
     /// appropriate `Request` to represent it.
-    pub fn from_reader<R: io::Read>(mut reader: R) -> Result<Request> {
+    ///
+    /// # Errors
+    /// This function will return an error if the request's reader blocks, or if the request cannot be parsed.
+    pub fn from_reader<R: io::Read>(mut reader: R) -> Result<Self> {
         let mut buffer = Vec::with_capacity(512);
-        let mut read_buf = [0u8; 512];
-
-        let mut req = loop {
-            let n = reader.read(&mut read_buf)?;
-            if n == 0 {
-                return Err(Error::ConnectionClosed);
-            }
-            buffer.extend_from_slice(&read_buf[..n]);
-            match http_parser::parse(std::mem::take(&mut buffer))? {
-                http_parser::Status::Complete(req) => break req,
-                http_parser::Status::Partial(b) => {
-                    let _ = mem::replace(&mut buffer, b);
+        loop {
+            match reader.read_to_end(&mut buffer) {
+                Ok(_) => break,
+                Err(ref e) if e.kind() == io::ErrorKind::WouldBlock => {
+                    if !buffer.is_empty() {
+                        break;
+                    }
                 }
-            }
-        };
+                Err(e) => return Err(Error::IO(e)),
+            };
+        }
 
-        if let Some(size) = req.header("Content-Length") {
-            let size = size.parse().unwrap_or(0);
-            let start = req.body.0;
+        let mut req = http_parser::parse(std::mem::take(&mut buffer))?;
 
-            while req.buffer[start..].len() < size {
-                let n = reader.read(&mut read_buf)?;
-                if n == 0 {
-                    break;
-                }
-                req.buffer.extend_from_slice(&read_buf[..n]);
-            }
-            req.body.1 = req.body.0 + size;
+        if req.header("Content-Length").is_some() {
             req.parse_form();
         }
 
         #[cfg(feature = "cookies")]
         {
             if let Some(cookie) = req.header("Cookie") {
-                let cookie = Cookie::parse(cookie).map_err(|e| Error::Other(e.to_string()))?;
-                let name = cookie.name().to_owned();
-                let val = util::percent_decode(cookie.value()).unwrap();
-                req.cookies.push((name, val));
+                if let Ok(cookies) = Cookie::parse(&cookie.into_owned()) {
+                    for cookie in cookies {
+                        req.cookies.push((
+                            cookie.get_name().to_string(),
+                            cookie.get_value().to_string(),
+                        ));
+                    }
+                }
```

Comment:
Old cookie crate was for parsing singular cookies, not the entire header at once